### PR TITLE
Adding nested MC sampling

### LINKED ISSUE #20130: Sampling a nested loop of parameters in STM
## Reason

There has been a recent request for STM to support a Wilk's sampling methodology, some information can be fount [here](https://en.wikipedia.org/wiki/Wilks%27_theorem). Basically, this entails having an outer and inner loop of parameter sampling. Parameters associated with the outer loop are only changed after each iteration of the inner loop, where those parameters are changed every time.

## Design

Although, it would be relatively straightforward to implement this sampling scheme directly, I find it easy to generalize it with an arbitrary number of nested loops. I'll call this object "NestedMonteCarlo", where the user can define any number of nested loops, with any number of columns and rows per loop.

## Impact

The goal is just to provide the user with the sampling  mechanics and all the parallelism along with it. Right now, it's not necessary to do all the fancy statistics along with it; we will leave that to the user offline until otherwise instructed.

### PR DESCRIPTION
**Pull request by zachmprince**
Closes #20130

### CODE REVIEW

File: modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      computeSample(target_row, j);
+    curr_row = target_row + 1;
+  }
+  restoreGeneratorState();
+}
+
+Real
+NestedMonteCarloSampler::computeSample(dof_id_type row_index, dof_id_type col_index)
+{
+  const Real rn = getRand();
```
Comment (loganharbour) [inline modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:88]: The random generator advancing on every sample is intentional, correct? Even though it's not always used? I assume this is needed for determinism?

File: modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  if (row_index % _loop_mod[_loop_index[col_index]] == 0)
```
Comment (loganharbour) [inline modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:89]: Is there any way to remove the double indexing here?

**Review by loganharbour — COMMENTED** [review]
Looks good to me. Just a few questions.

File: modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  for (const auto & mod : _loop_mod)
+  {
+    if (getLocalRowBegin() % mod == 0)
+      break;
+
+    const dof_id_type target_row = std::floor(getLocalRowBegin() / mod) * mod;
+    advanceGenerators((target_row - curr_row) * getNumberOfCols());
+    for (const auto & j : make_range(getNumberOfCols()))
+      computeSample(target_row, j);
+    curr_row = target_row + 1;
```
Comment (loganharbour) [inline modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:80]: FYI this is never hit

File: modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C (hunk 1 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:88]: Yes. The base class assumes that number of random numbers needed is `_n_rows * _n_cols`. So when a full sampling is complete the generators need to advance exactly that much.

File: modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C (hunk 2 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:89]: Not easily. I need something to map the column index to the loop that the column is a part of. I can create a separate variabel to hold the loop index above.

File: modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C (hunk 3 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:80]: Yeah, I possibly need to modify the test. Does the coverage look at what code is hit by all processes in parallel?

Comment (loganharbour) [reply modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:80]: The entries used for coverage from modules come from these three recipes:

- Modules: https://civet.inl.gov/job/941209/
- Modules Parallel: https://civet.inl.gov/job/941210/
- Modules Recover: https://civet.inl.gov/job/941212/

For stochastic tools it looks like we have:

- `-p 1`
- `-p 2`
- `-p 2 --distributed-mesh`
- `--n-threads 2`
- `-p 2 --n-threads 2 --distributed-mesh`

Comment (zachmprince) [reply modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:80]: Okay, so I guess it gathers the coverage from processors other than root too. I added a test to make sure that gets hit.

Comment (zachmprince) [reply modules/stochastic_tools/src/samplers/NestedMonteCarloSampler.C:80]: Coverage looks good now. Thanks for the tip.

**Review by loganharbour — APPROVED** [review]
Merge as you see fit

**Pull request merged**